exp/image/colour.d

```diff
@@ -0,0 +1,695 @@
+alias Colour!("rgba", ushort) RGBA16;
+
+alias Colour!("l", ubyte) L8;
+alias Colour!("l", ushort) L16;
+alias Colour!("la", ubyte) LA;
+alias Colour!("la", ushort) LA16;
+
+alias Colour!("rg", byte) UV8;
+
+alias Colour!("xyz", float, ColourSpace.XYZ) XYZ;
+
+
+// TODO: represent packed colours, eg R5G6B5, etc
+struct PackedColour(string format, ColourSpace colourSpace = ColourSpace.sRGB)
+{
+	// TODO: we'll try and fabricate the packing algorithm based on the format string
+	//...
+
+	void pack(C)(C colour);
+	C unpack(C)();
+}
+
+// std.conv style to! function, which performs almost all conversion operations
+To to(To, From)(From colour) if(isColour!To && isColour!From)
+{
+	static if(is(To == Colour!(cmpT, TyT, csT), string cmpT, TyT, ColourSpace csT) && is(From == Colour!(cmpF, TyF, csF), string cmpF, TyF, ColourSpace csF))
+	{
+		auto src = colour.tristimulusWithAlpha;
+
+		static if(csF == csT)
+		{
+			// colour space is the same, just do type conversion
+			return To(convert!TyT(src[0]), convert!TyT(src[1]), convert!TyT(src[2]), convert!TyT(src[3]));
+		}
+		else
+		{
+			static if(false && TyF.sizeof <= 2 && TyT.sizeof <= 2)
+			{
+				// <= 16bit type conversion should use a look-up table
+
+				// TODO: this is blocked by '^^' not working in CTFE!
+			}
+			else
+			{
+				// full colour space conversion
+
+				// unpack the working values
+				alias WorkType = WorkingType!(TyF, TyT);
+				WorkType a = convert!WorkType(src[0]);
+				WorkType b = convert!WorkType(src[1]);
+				WorkType c = convert!WorkType(src[2]);
+
+				// perform the colour space conversion
+				auto r = convertColourSpace!(csF, csT)(a, b, c);
+
+				// convert and return the output
+				return To(convert!TyT(a), convert!TyT(b), convert!TyT(c), convert!TyT(src[3]));
+			}
+		}
+	}
+}
+
+
+// helper functions
+
+T sRGBToLinear(T)(T s) pure if(isFloatingPoint!T)
+{
+	if(s <= T(0.04045))
+		return s / T(12.92);
+	else
+		return ((s + T(0.055)) / T(1.055))^^T(2.4);
+}
+T gammaToLinear(T)(T s) pure if(isFloatingPoint!T)
+{
+	return s^^T(2.2);
+}
+T adobeToLinear(T)(T s) pure if(isFloatingPoint!T)
+{
+	return s^^T(2.19921875);
+}
+
+T linearTosRGB(T)(T s) pure if(isFloatingPoint!T)
+{
+	if(s <= T(0.0031308))
+		return T(12.92) * s;
+	else
+		return T(1.055) * s^^T(1.0/2.4) - T(0.055);
+}
+T linearToGamma(T)(T s) pure if(isFloatingPoint!T)
+{
+	return s^^T(1.0/2.2);
+}
+T linearToAdobe(T)(T s) pure if(isFloatingPoint!T)
+{
+	return s^^T(1.0/2.19921875);
+}
+
+T toGrayscale(ColourSpace colourSpace = ColourSpace.sRGB, T)(T r, T g, T b) pure if(isFloatingPoint!T)
+{
+/*
+	TODO: do i have this function right? am i in the wrong space?
+
+	// i think this is the linear conversion coefficients
+	wiki = 0.299r + 0.587g + 0.114b
+	ps: (0.3)+(0.59)+(0.11)
+
+	// but gimp source seems to use this one... perhaps this performs an sRGB space estimate, or is THIS the linear one?
+	gimp: 0.21 R + 0.72 G + 0.07 B
+	another = 0.2126 R + 0.7152 G + 0.0722 B   (apparently in linear space?)
```

Comment:
Yeah, I'll dig out the truth. I was focusing more on the structure of the thing, and whether I like design ideas like fabricating colour components from elements in a string... not entirely sure about that, but it's kinda convenient.